distributed/scheduler.py

```diff
@@ -3792,25 +3801,162 @@ async def add_nanny(self, comm):
+                    ts.exception_blame = ts
+                    parent.tasks[key] = ts
+
+                self.client_desires_keys(keys=keys, client=client)
+                return
+            finally:
+                _current_client.reset(tok)
+        else:
+            graph_ = graph
+
+        if annotations is None:
+            annotations = {}
+
+        workers = workers or annotations.pop("workers", None)
+        allow_other_workers = allow_other_workers or annotations.pop(
+            "allow_other_workers", None
+        )
+        if retries is None:
+            retries = annotations.pop("retries", None)
+        resources = resources or annotations.pop("resources", None)
+        user_priority = annotations.pop("priority", user_priority)
+
+        if isinstance(workers, (str, Number)):
+            workers = [workers]
+        if isinstance(workers, (tuple, set)):
+            workers = list(workers)
+        if isinstance(workers, list):
+            restrictions = workers
+        elif workers is None:
+            restrictions = []
+        else:
+            raise TypeError("Workers must be a list or set of workers or None")
+
+        dsk = dict(graph_)
+
+        if allow_other_workers:
+            loose_restrictions = set(dsk)
+        else:
+            loose_restrictions = set()
+
+        from distributed.utils_comm import unpack_remotedata
+
+        dependencies, dependents = get_deps(dsk)
+
+        # Remove `Future` objects from graph and note any future     dependencies
+        dsk2 = {}
+        fut_deps = {}
+        for k, v in dsk.items():
+            dsk2[k], futs = unpack_remotedata(v, byte_keys=True)
+            if futs:
+                fut_deps[k] = futs
+        dsk = dsk2
+
+        # - Add in deps for any tasks that depend on futures
+        for k, futures in fut_deps.items():
+            dependencies[k].update(f.key for f in futures)
+
+        pre_stringify = set(dsk)
+        dsk = {stringify(k): stringify(v, exclusive=graph_) for k, v in dsk.items()}
+
+        def process(x, keys=dsk, string_keys=pre_stringify):
+            if callable(x):
+                return {stringify(k): x(k) for k in keys}
+            elif isinstance(x, (int, dict, tuple, set, list)):
+                return {stringify(k): x for k in string_keys or map(stringify, keys)}
+            elif isinstance(x, dict):
+                return keymap(stringify, x)
+            raise TypeError()
+
+        if annotations:
+            annotations = process(annotations)
+        if retries:
+            retries = process(retries)
+        if resources:
+            resources = process(resources)
+        if user_priority:
+            user_priority = process(user_priority)
+        if restrictions:
+            _restrictions = process(restrictions)
+        else:
+            _restrictions = {}
+
+        for layer in graph_.layers.values():
+            if layer.annotations and "retries" in layer.annotations:
```

Comment:
Nit: we are checking if `layer.annotations` exists a lot in this block. We could check once at the top:
```python
if not layer.annotations:
    continue
````